program/src/instruction.rs

```diff
@@ -258,50 +264,72 @@ where
     })
 }
 
+/// Create `VerifyTransferSignature` instruction
+pub fn verify_transfer_signature(
+    program_id: &Pubkey,
+    verified_messages: &Pubkey,
+    reward_manager: &Pubkey,
+    sender: &Pubkey,
+    funder: &Pubkey,
+) -> Result<Instruction, ProgramError> {
+    let data = Instructions::VerifyTransferSignature.try_to_vec()?;
+
+    let accounts = vec![
+        AccountMeta::new(*verified_messages, false),
+        AccountMeta::new_readonly(*reward_manager, false),
+        AccountMeta::new_readonly(*sender, false),
+        AccountMeta::new(*funder, true),
+        AccountMeta::new_readonly(sysvar::instructions::id(), false),
+        AccountMeta::new_readonly(system_program::id(), false),
+    ];
+
+    Ok(Instruction {
+        program_id: *program_id,
+        accounts,
+        data,
+    })
+}
+
 /// Create `Transfer` instruction
-pub fn transfer<I>(
+#[allow(clippy::too_many_arguments)]
+pub fn transfer(
     program_id: &Pubkey,
+    verified_messages: &Pubkey,
     reward_manager: &Pubkey,
-    recipient: &Pubkey,
-    vault_token_account: &Pubkey,
+    reward_token_source: &Pubkey,
+    reward_token_recipient: &Pubkey,
     bot_oracle: &Pubkey,
-    funder: &Pubkey,
-    senders: I,
-    params: Transfer,
-) -> Result<Instruction, ProgramError>
-where
-    I: IntoIterator<Item = Pubkey>,
-{
-    let data = Instructions::Transfer(Transfer {
-        amount: params.amount,
-        id: params.id.clone(),
-        eth_recipient: params.eth_recipient,
+    payer: &Pubkey,
+    amount: u64,
+    id: String,
+    eth_recipient: [u8; 20],
+) -> Result<Instruction, ProgramError> {
+    let data = Instructions::Transfer(TransferArgs {
+        amount,
+        id: id.clone(),
+        eth_recipient,
     })
     .try_to_vec()?;
 
-    let transfer_acc_to_create = get_address_pair(
+    let transfer_account = get_address_pair(
         program_id,
         reward_manager,
-        [TRANSFER_SEED_PREFIX.as_bytes().as_ref(), params.id.as_ref()].concat(),
+        [TRANSFER_SEED_PREFIX.as_bytes().as_ref(), id.as_ref()].concat(),
     )?;
 
-    let mut accounts = vec![
+    let accounts = vec![
+        AccountMeta::new(*verified_messages, false),
         AccountMeta::new_readonly(*reward_manager, false),
-        AccountMeta::new_readonly(transfer_acc_to_create.base.address, false),
```

Comment:
There are no base accounts anymore, please rename to the reward manager authority.